src/resolve/resolved-dns-answer.c

```diff
@@ -181,11 +181,19 @@ int dns_answer_add(
 
         exist = ordered_set_get(a->items, &tmp);
         if (exist) {
-                /* There's already an RR of the same RRset in place! Let's see if the TTLs more or less
-                 * match. We don't really care if they match precisely, but we do care whether one is 0 and
-                 * the other is not. See RFC 2181, Section 5.2. */
-                if ((rr->ttl == 0) != (exist->rr->ttl == 0))
+                /* There's already an RR of the same RRset in place! Let's see if the TTLs more or
+                 * less match. RFC 2181, Section 5.2 suggests clients should reject RRsets
+                 * containing RRs with differing TTLs. We are more tolerant of this situation except
+                 * if one RR has a zero TTL and the other a nonzero TTL. In mDNS, zero TTLs are
+                 * special, so we must error in that case. */
+                if ((rr->ttl == 0) != (exist->rr->ttl == 0) &&
+                    ((exist->flags & DNS_ANSWER_REFUSE_TTL_NO_MATCH) ||
+                     (flags & DNS_ANSWER_REFUSE_TTL_NO_MATCH))) {
+                        log_debug("Refusing to merge RRs with zero TTL and nonzero TTL: %s %s",
+                                  dns_resource_record_to_string(rr),
+                                  dns_resource_record_to_string(exist->rr));
                         return -EINVAL;
```

Comment:
I think we should still log something even if we do the merging. Maybe something like this:

```c
if ((rr->ttl == 0) != (exist->rr->ttl == 0)) {
        if ((exist->flags | flags) & DNS_ANSWER_REFUSE_TTL_NOT_MATCH))
                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Refusing to merge RRs with zero TTL and non-zero TTL: %s vs. %s",
                                dns_resource_record_to_string(rr),
                                dns_resource_record_to_string(exist->rr));

        log_debug("Merging RRs with zero TTL and non-zero TTL (not RFC 2181/5.2 compliant): %s vs. %s",
                        dns_resource_record_to_string(rr),
                        dns_resource_record_to_string(exist->rr));
}
```

Reply:
Will do. I wasn't sure if we should log for all cases where the ttls match, which is what RFC 2181 asks for, or just the zero/non-zero case.